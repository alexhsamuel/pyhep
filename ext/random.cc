//----------------------------------------------------------------------
//
// random.cc
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <cstddef>
#include <sys/time.h>

#include "random.hh"

//----------------------------------------------------------------------
// constants
//----------------------------------------------------------------------

namespace {

// Constants for the first congruential generator.
const static long m1 = 2147483563;
const static long a1 = 40014;
const static long q1 = 53668;
const static long r1 = 12211;

// Constants for the second congruential generator.
const static long m2 = 2147483399;
const static long a2 = 40692;
const static long q2 = 52774;
const static long r2 = 3791;

//----------------------------------------------------------------------
// helper functions
//----------------------------------------------------------------------

inline float
min(float v1,
    float v2)
{
  return (v1 < v2) ? v1 : v2;
}


}  // anonymous namespace

//----------------------------------------------------------------------
// methods
//----------------------------------------------------------------------

ShuffledLEcuyerRandom::ShuffledLEcuyerRandom(long seed)
{
  initialize(seed);
}


void
ShuffledLEcuyerRandom::initialize(long seed)
{
  // If no seed was specified, construct one from the system time.
  if (seed == 0) {
    struct timeval now;
    int result = gettimeofday(&now, NULL);
    assert(result == 0);
    seed = now.tv_sec ^ now.tv_usec;
  }

  // Initialize.
  assert(seed > 0);
  state1_ = seed;
  state2_ = seed;

  // Load the shuffle table (after eight warm-ups).
  for (long j = table_size_ + 7; j >= 0; j--) {
    long k = state1_ / q1;
    state1_ = a1 * (state1_ - k * q1) - k * r1;
    if (state1_ < 0)
      state1_ += m1;
    if (j < table_size_)
      shuffle_table_[j] = state1_;
  }
  last_ = shuffle_table_[0];
}


float
ShuffledLEcuyerRandom::random()
{
  long k;
  static float inv_m1 = 1.0 / m1;

  // Next value from the first congruential generator.
  k = state1_ / q1;
  state1_ = a1 * (state1_ - k * q1) - k * r1;
  if (state1_ < 0)
    state1_ += m1;

  // Next value from the second congruential generator.
  k = state2_ / q2;
  state2_ = a2 * (state2_ - k * q2) - k * r2;
  if (state2_ < 0)
    state2_ += m2;

  // Combine and shuffle.
  int j = last_ / (1 + (m1 - 1) / table_size_);
  last_ = shuffle_table_[j] - state2_;
  shuffle_table_[j] = state1_;
  if (last_ < 1)
    last_ += m1 - 1;

  // Make sure we never return 1.0.
  return min(last_ * inv_m1, 1.0 - 1.2e-7);
}


