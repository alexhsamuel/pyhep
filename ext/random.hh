//----------------------------------------------------------------------
//
// random.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

#ifndef __RANDOM_HH__
#define __RANDOM_HH__

//----------------------------------------------------------------------
// classes
//------------------------------------------------------------------------

/* Shuffled L'Ecuyer random number generator.

   Generator for uniform deviates based on 

     L'Ecuyer, P. 1988, Communications of the ACM, vol. 31, pp. 742-774

   as described and implemented as the 'ran2' function in

     Press, William H. et al., Numerical Recipes in C, Cambridge
     University Press, 1988, p. 282.

   This generator is based on two multiplicative congruential generators
   with factors 'a1' and 'a2' and moduli 'm1' and 'm2, with
   multiplication performed with Schrage's algorith.  The values from
   the two sequences are combined, and the sequence is shuffled using a
   table of 'tab_size'.  

   The period of the generator is about 2.3 x 10^18.

*/
class ShuffledLEcuyerRandom
{
public:

  /* Create a new generator.

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  ShuffledLEcuyerRandom(long seed=0);
  
  /* Initialize the generator. 

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  void initialize(long seed=0);

  /* Return a uniform deviate between 0 inclusive and 1 exclusive.  */
  float random();

private:

  // Forbid copies.
  ShuffledLEcuyerRandom(const ShuffledLEcuyerRandom&);

  // Shuffle table size.
  const static int table_size_ = 32;

  // State variables for the congruential generators.
  long state1_;
  long state2_;

  // State of the combined generator; used for shuffling.
  long last_;

  // The shuffle table.
  long shuffle_table_[table_size_];

};

//----------------------------------------------------------------------

#endif  // #ifndef __RANDOM_HH__
