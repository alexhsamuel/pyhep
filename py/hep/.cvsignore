*.pyc
*.pyo
config.py
*-install
pdt.pickle
