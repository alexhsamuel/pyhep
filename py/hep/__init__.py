#-----------------------------------------------------------------------
#
# module hep
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Module for general-purpose code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

import os
import signal
import traceback

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TimeoutException(Exception):
    """A timeout occurred."""

    pass



class NotSupportedError(RuntimeError):
    """A feature is not supported."""

    pass



class SyntaxError(Exception):

    pass



#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Token:

    pass



class Timeout:

    def __timeout_handler(self, signal_number, frame):
        assert signal_number == signal.SIGALRM
        signal.signal(signal.SIGALRM, self.__old_handler)
        self.__active = 0
        raise TimeoutException, self.__message


    def __init__(self, time, message):
        self.__message = message
        self.__old_handler = \
            signal.signal(signal.SIGALRM, self.__timeout_handler)
        self.__active = 1
        signal.alarm(time)


    def cancel(self):
        if not self.__active:
            return
        signal.alarm(0)
        signal.signal(signal.SIGALRM, signal.SIG_DFL)
        self.__active = 0



class MappingAsObject:
    """Implement the object attribute protocol using a mapping.

    A 'MappingAsObject' wraps an object satifying the mapping protocol.
    Attribute references are converted to item gets and sets in the
    underlying mapping."""

    def __init__(self, mapping):
        """Wrap 'mapping' in the object procotol."""
        
        self.__dict__["_MappingAsObject__mapping"] = mapping


    _mapping = property(lambda self: self.__mapping)
    """Read-only access to the underlying mapping."""


    def __getattr__(self, name):
        try:
            return self.__mapping[name]
        except KeyError:
            raise AttributeError, name


    def __setattr__(self, name, value):
        try:
            self.__mapping[name] = value
        except KeyError, exception:
            raise AttributeError, str(exception)



class NrBreitWigner:
    """A non-resonant Breit-Wigner line shape amplitude factor."""

    def __init__(self, mass, width):
        """Construct a non-relativistic Breit-Wigner function.

        'mass' -- The resonance mass.
        
        'width' -- The resonance width."""

        self.__mass = mass
        self.__width = width


    def __call__(self, s):
        """Return the amplitude at energy-squared 's'."""

        mass = self.__mass
        return 1 / (s - complex(mass * mass, mass * self.__width))



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

try:
    enumerate = __builtins__.enumerate
except AttributeError:
    def enumerate(iterable):
        count = 0
        for item in iter(iterable):
            yield count, item
            count += 1


try:
    sorted = __builtins__.sorted
except AttributeError:
    def sorted(iterable, cmp=cmp):
        result = list(iterable)
        result.sort(cmp)
        return result


try:
    reversed = __builtins__.reversed
except AttributeError:
    def reversed(iterable):
        result = list(iterable)
        result.reverse()
        return result
    

def sort(sequence, sort_fn=None):
    """Return a copy of 'sequence', sorted by 'sort_fn'.

    'sort_fn' -- The comparison function to use.  If 'None', the default
    (built-in 'cmp') is used."""

    sequence = list(sequence)
    if sort_fn is None:
        sort_fn = cmp
    sequence.sort(sort_fn)
    return sequence


def binarySearch(sequence, value, compare=cmp):
    """Find 'value' in sorted 'sequence' by binary search.

    'sequence' -- A sequence of values.  It is assumed to be sorted by
    'compare'.

    'compare' -- A comparison function.

    returns -- The index of the largest element in 'sequence' smaller
    than 'value', or -1 if there is no such element."""

    if len(sequence) == 0:
        return -1
    if compare(value, sequence[0]) < 0:
        return -1
    if compare(value, sequence[-1]) > 0:
        return len(squence) - 1

    i0 = 0
    i2 = len(sequence) - 1
    while True:
        if i2 <= i0 + 1:
            return i2 - 1
        i1 = (i0 + i2) // 2
        if compare(value, sequence[i1]) < 0:
            i2 = i1
        else:
            i0 = i1


def mapFromKeys(key_sequence, value_function):
    """Construct a map from a sequence of keys.

    A map is created whose keys are the elements of 'key_sequence'.  For
    each key, the corresponding value is constructed from the key using
    'value_function'.  If 'sequence' contains a particular key more than
    once, only one is placed in the map (which one is undefined).

    'key_sequence' -- A sequence of key values.

    'value_function' -- A callable for determining the value for each
    element of the sequence.  It is invoked for each key, passing the
    kay as the single argument."""
    
    result = {}
    for key in key_sequence:
        result[key] = value_function(key)
    return result


def mapFromValues(key_function, value_sequence):
    """Construct a map from a sequence of values.

    A map is created whose values are the elements of 'value_sequence'.
    For each value, the corresponding key is constructed from the value
    using 'key_function'.  If 'key_function' assigns the same key to
    more than one element of the sequence, only one element is actually
    used in the map (which one is undefined).

    'key_function' -- A callable for determining the key for each
    element of the sequence.  It is invoked for each value, passing
    the value as the single argument.

    'value_sequence' -- A sequence of values."""
    
    result = {}
    for value in value_sequence:
        result[key_function(value)] = value
    return result


def invertMap(map):
    """Produce a dictionary by inverting keys and values of 'map'."""

    return dict([ (value, key) for key, value in map.items() ])


def remove(dict, key):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    returns -- The value corresponding to 'key' in 'dict'.  

    raises -- 'KeyError' if 'dict' has no key 'key'."""
    
    value = dict[key]
    del dict[key]
    return value


def popkey(map, key, default=None):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    'default' -- The value to return if 'key' is not in 'map'.

    returns -- The value corresponding to 'key' in 'dict', or 'default'
    if none."""

    if key in map:
        result = map[key]
        del map[key]
        return result
    else:
        return default


def getHostName():
    """Return the name of this host."""

    import socket

    try:
        return socket.gethostbyname_ex(socket.gethostname())[0]
    except socket.error:
        return "(unknown)"


def isMap(object):
    """Determine whether an object satisfies the mapping protocol.

    returns -- True if 'object' satisfies the map protocol as
    specified.""" 

    # FIXME.  There probably is a more consistent way to do this.

    # If the object is an instance of the standard 'dict' class (or a
    # subclass of it), it must satisfy the mapping protocol.
    if isinstance(object, dict):
        return True
    # Check for the required methods.
    for method_name in ("get", "__getitem__", "__len__", ):
        if not hasattr(object, method_name):
            return False
    # The object checks out.
    return 1


def mkdirRecursively(path):
    """Create a directory, including parent directories if necessary."""

    parent, name = os.path.split(path)
    if parent != "" and not os.path.isdir(parent):
        mkdirRecursively(parent)
    os.mkdir(path)


def rmRecursively(path):
    """Remove the file or directory at 'path'.

    If 'path' is a directory, its contents are removed first."""

    if os.path.isdir(path):
        # Construct a list of contents of the directory.
        contents = map(lambda e, path=path: os.path.join(path, e),
                       os.listdir(path))
        # Remove them all.
        map(rmRecursively, contents)
        # Remove the directory itself.
        os.rmdir(path)
    else:
        os.unlink(path)


