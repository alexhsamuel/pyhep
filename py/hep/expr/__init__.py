#-----------------------------------------------------------------------
#
# module hep.expr
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Compiled arithmetical expressions."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   classes import *
import compiler
from   compiler import isCompiledExpression
import expr_parser
from   hep.bool import *
import hep.ext
import inspect
import op
from   op import optimize, substitute, substituteExpressions
from   op import setTypes, setTypesFrom, setTypesFixed
import re
from   symbols import substituteConstants

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def isExpression(object):
    """Return true if 'object' is an expression object."""

    return isCompiledExpression(object) \
           or isinstance(object, Expression)


def asExpression(arg, compile=False, types=None, types_from=None, **names):
    """Interpret 'arg' as an expression.

    For maximum convenience, this function is called by (most)
    user-visible functions which take expression arguments.

    'compile' -- If true, returns a compiled expression.

    '**names' -- Values to substitute for symbol names in the
    expression.

    returns -- If 'arg' is an expression, itself; or 'arg' converted to
    an expression.

    raises -- 'TypeError' if 'arg' is not an expression and cannot be
    converted to one."""

    # If it's already an expression, just use it.
    if isExpression(arg):
        expr = arg

    # If it's a string, parse it.
    elif isinstance(arg, str):
        try:
            expr = parse(arg)
        except expr_parser.ParseError, exception:
            # Make parse exceptions more useful by including the parsed
            # string. 
            raise expr_parser.ParseError, \
                  "%s, while parsing \"%s\"" % (exception, arg)

    # If it's a function, fish out its argument names, make symbols out
    # of these, and wrap it all in a call expression.
    elif callable(arg):
        if inspect.isfunction(arg):
            # An ordinary function.
            symbol_names = inspect.getargspec(arg)[0]
        elif inspect.ismethod(arg):
            # A class method.
            symbol_names = inspect.getargspec(arg.im_func)[0]
            # If it's a bound method, skip the first ('self') argument.
            if arg.im_self is not None:
                symbol_names = symbol_names[1 :]
        elif True:
            symbol_names = ("x", )
        else:
            # Don't know how to get the symbol names.
            raise ValueError, \
                  "don't know how to get argument names for '%r'" \
                  % arg
        # Construct the expression.
        subexprs = map(Symbol, symbol_names)
        expr = Call(Constant(arg), subexprs)

    # If it's numerical, wrap it in a constant.
    elif type(arg) in (int, float, long):
        expr = Constant(arg)

    # Don't know how to handle other things.
    else:
        raise TypeError, "argument must be an expression"

    # Substitute names.
    expr = substitute(expr, **names)
    # Set types.
    if types_from is not None:
        expr = setTypesFrom(expr, **types_from)
    if types is not None:
        expr = setTypes(expr, **types)
    # Compile it, if requested.
    if compile:
        expr = compiler.compile(expr)
    # All done.
    return expr


def parse(expression_string, **names):
    expression = expr_parser.parse(expression_string)
    expression = substitute(expression, **names)
    return expression


def compile(expression, default_type=None, **types):
    """Compile 'expression'.

    'expression' -- """

    if isCompiledExpression(expression):
        # It is already compiled.  
        return expression
    expression = asExpression(expression)
    
    if default_type is not None:
        # Set default symbol types.
        expression = op.setTypesFixed(expression, None, default_type)
    if len(types) != 0:
        # Set specific symbol types.
        expression = op.setTypes(expression, **types)

    # Do the compilation.
    expression = op.optimize(expression)
    return compiler.compile(expression)


def parseFile(lines, symbols={}, expressions=None):
    if expressions is None:
        expressions = {}

    name_regex = re.compile(r"(\w+)\s*=(\s|\Z)")

    name = None
    text = None
    def finish():
        if name is None:
            return
        expression = substituteExpressions(parse(text), expressions)
        expression = substitute(expression, **symbols)
        expressions[name] = expression

    for line in lines:
        line = line.strip()
        # Trim off comments.
        if "#" in line:
            line = line[:line.index("#")]
        # Skip blank lines.
        if line.strip() == "":
            continue

        # Is this the start of a new definition?
        match = name_regex.match(line)
        if match is not None:
            finish()
            name = match.group(1)
            text = line[match.end():]

        elif name is not None:
            text += " " + line

        else:
            raise RuntimeError, "syntax error"
        
    finish()
    return expressions


def _getSymbolNames(expr, result):
    if isinstance(expr, Symbol):
        name = expr.symbol_name
        result[name] = None
    for subexpr in expr.subexprs:
        _getSymbolNames(subexpr, result)


def getSymbolNames(expr):
    """Return a list of names of symbols in 'expr'."""
    
    result = {}
    _getSymbolNames(expr, result)
    return result.keys()
