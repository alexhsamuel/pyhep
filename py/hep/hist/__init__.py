#-----------------------------------------------------------------------
#
# module hep.hist
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Histograms.

The following error models are supported:

  'explicit' -- Asymmetrical errors may be specified for each bin using
  the 'setError' method.  No errors are computed automatically.

  'gaussian' -- The symmetrical error on the value of each bin is
  computed assuming a Gaussian distribution.

  'none' -- Bins have zero errors.

  'poisson' -- The asymmetrical error on the value of each bin is computed
  assuming a Poisson distribution.

  'quadrature' -- The symmetrical error on each bin is square root of
  the sum of squares of weights accumulated to that bin.

Error values are always of type 'float'.
"""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

from   axis import Axis, BinnedAxis, AxisIterator, AxesIterator
import copy
import copy_reg
from   function import Function1D, isFunction, SampledFunction1D, \
                       sampleFunction1D
from   function import Graph, ContourGraph, isGraph, \
                       makeContourGraphFromHist
from   hep import ext
from   hep.bool import *
from   histogram import Histogram, Histogram1D, isHistogram
from   scatter import isScatter, Scatter
import sys
from   util import *

