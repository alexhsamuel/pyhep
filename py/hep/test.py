#-----------------------------------------------------------------------
#
# module hep.test
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Regression testing support."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.bool import *
import sys

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TestFailure(Exception):
    pass



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def assert_(expression):
    if not expression:
        raise TestFailure, "assertion failed"


def compare(value, expected, precision=None):
    if type(value) in (list, tuple) and type(expected) in (list, tuple):
        if len(value) != len(expected):
            result = False
        else:
            try:
                result = reduce(
                    lambda r1, r2: r1 and r2,
                    map(lambda (v, e): compare(v, e, precision),
                        zip(value, expected)),
                    True)
            except TestFailure:
                result = False

    elif precision is not None:
        result = (abs(value - expected) <= precision)

    else:
        result = (value == expected)

    if result:
        return True
    else:
        raise TestFailure, \
              "expected %s; got %s" % (repr(expected), repr(value))


def compareSequence(value, expected, ordered=True):
    value = list(value)
    expected = list(expected)
    if not ordered:
        value.sort()
        expected.sort()
    return compare(value, expected)


