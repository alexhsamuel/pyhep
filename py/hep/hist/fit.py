#--------------------------------------------------*- coding: Latin1 -*-
#
# fit.py
#
# Copyright (C) 2004 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep import enumerate
from   hep.bool import *
import hep.expr
import hep.expr.op
import hep.fn
import hep.hist
from   hep.num import sum, product, factorial
from   math import hypot, sqrt, log
import sys

#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

# FIXME: 'NaiveChiSquare' and 'naiveChiSquareFit' are deprecated, and
# are only left here only until another multidimensional fitter is
# available.

class NaiveChiSquare:

    def __init__(self, function, sample_vars, samples):
        self.function = function
        self.sample_vars = tuple(sample_vars)
        self.samples = samples


    def __call__(self, **parameters):
        symbols = dict(parameters)

        result = 0
        for x, y, e in self.samples:
            # FIXME:
            if e == 0:
                continue
            for var, value in zip(self.sample_vars, x):
                symbols[var] = value
            f = self.function(**symbols)
            result += ((f - y) / e) ** 2

        return result



def naiveChiSquareFit(histogram, function, variable_names, parameters):
    import hep.cernlib.minuit

    function = hep.expr.asExpression(function)
    samples = [ (hep.hist.getBinCenter(histogram.axes, bin),
                 histogram.getBinContent(bin),
                 histogram.getBinError(bin))
                for bin in hep.hist.AxesIterator(histogram.axes) ]
    chi_sq = NaiveChiSquare(function, variable_names, samples)
    result = hep.cernlib.minuit.minimize(chi_sq, *parameters)
    result.expression = hep.expr.substitute(function, **result.values)
    return result


#-----------------------------------------------------------------------

class ChiSquare1D:
    """Function object to compute one-dimensional chi-square."""

    def __init__(self, function, sample_var, samples):
        self.function = function
        self.sample_var = sample_var
        self.samples = samples
        self.warned = False


    def __call__(self, **parameters):
        # FIXME: We could save some evaluations on adjacent bins here.
        def eval(x, sample_var=self.sample_var, function=self.function):
            parameters[sample_var] = x
            return function(**parameters)

        # Substitute parameters.
        function = hep.expr.substitute(self.function, **parameters)
        # Compile the function, using 'float' as the type for the
        # independent variable.
        function = hep.expr.compile(function, float)

        eval = lambda v: function.evaluate({ self.sample_var: v })

        chi_sq = 0
        for (x0, x1), y in self.samples:
            # We can't handle non-positive bin values, since we can't
            # compute gaussian errors for them.
            if y <= 0:
                if not self.warned:
                    print >> sys.stderr, \
                      "Warning: non-positive bin values ignored in chi-square"
                    self.warned = True
                continue
            # Do Simpson's rule integration.
            f = (x1 - x0) \
                * (eval(x0) + 4 * eval((x0 + x1) / 2) + eval(x1)) / 6
            # Update the chi-square.
            chi_sq += ((f - y) ** 2) / y

        return chi_sq



def chiSquareFit1D(histogram, function, variable_name, parameters,
                   min_count=7, range=None):
    """Perform a one-dimensional chi-square fit of a PDF to a histogram.
    
    Performs a chi-square fit of PDF 'function' to a one-dimensional
    histogram.  The chi-square is computed using Simpson's rule
    (three-point gaussian quadrature) to integrate the PDF for
    comparison with histogram bins.  Bin errors in 'histogram' are
    ignored; 1/sqrt(N) gaussian couning errors are always used.
    Therefore, histogram contents should be positive integers (though
    zero and small bins are handled; see the 'min_count' parameter
    below).

    'histogram' -- The histogram to fit to.

    'function' -- The PDF.  May be an 'Expression' object or a
    callable.

    'variable_name' -- The name of the variable in 'function'
    corresponding to the dependent quantity in 'histogram' (i.e. the
    quantity along the histogram axis).

    'parameters' -- A sequence of parameters to fit.  Each may be simply
    a variable name in 'function', or a sequence '(name, initial_value,
    step_size, bound_low, bound_hi)'.

    'min_count' -- The minimum value for bins in the fit.  Bins which
    contain a smaller value than this are coalesced with neighboring
    bins until all bins contain at least this value.

    returns -- A 'Result' object."""

    # We'll use this for minimization.
    import hep.minimize
    import hep.cernlib.minuit

    axis = histogram.axis
    function = hep.expr.asExpression(function)
    parameters = hep.minimize.normalizeParameters(parameters)

    num_floating = sum([ 1 for p in parameters if p[2] is not None ])

    # Construct the list of bin samples, coalescing bins as necessary. 
    samples = []
    # Figure out which bins are included in the fit.
    if range is None:
        bins = hep.hist.AxisIterator(axis)
    else:
        lo, hi = range
        bins = xrange(axis(lo), axis(hi))
    # Loop over bins.
    x0 = None
    for bin in bins:
        range = axis.getBinRange(bin)
        # Are we carrying a bin whose contents don't satisfy the minimum
        # so far? 
        if x0 is None:
            # No.  Start a new bin
            x0, x1 = range
            value = histogram.getBinContent(bin)
        else:
            # Yes.  Make sure we are adjacent.
            assert x1 == range[0]
            # Coalesce this bin with the carried one.
            x1 = range[1]
            value += histogram.getBinContent(bin)
        # Doe the bin we're carrying have enough in it?
        if value >= min_count:
            # Yes it does.  Use it as a sample.
            samples.append(((x0, x1), value, ))
            x0 = None

    # If we're still carrying a bin, it must not have enough in it.  But
    # we have to do something with it.
    if x0 is not None:
        # Are there any samples at all so far?
        if len(samples) > 0:
            # Yes there are.  Coalesce the bin we're carrying with the
            # last one. 
            (last_x0, last_x1), last_value = samples.pop(-1)
            assert last_x1 == x0
            samples.append(((last_x0, x1), last_value + value, ))
        else:
            # No samples.  The whole histogram doesn't have enough in it
            # to satisfy the minimum criterion.  Oh well, just use one
            # sample. 
            samples.append(((x0, x1), value, ))

    # Construct the chi-square function.
    chi_sq = ChiSquare1D(function, variable_name, samples)
    # Minimize it.
    result = hep.cernlib.minuit.minimize(chi_sq, parameters)

    # Store the number of degrees of freedom.
    dof = len(samples) - num_floating
    result.degrees_of_freedom = dof
    # Compute the chi-square probability.
    result.chi_square_probability = \
        dof > 0 and hep.num.chiSquareCDF(result.minimum, dof) or None
                                 
    # Substitute the fit parameter values into 'function' and store
    # that. 
    result.expression = hep.expr.substitute(function, **result.values)

    # Construct a function suitable for plotting.  Scale the PDF by the
    # bin size, for sensible comparison to the histogram.
    bin_size = (axis.range[1] - axis.range[0]) / axis.number_of_bins
    scaled_expr = hep.expr.Multiply(
        result.expression, hep.expr.Constant(bin_size))
    notes = "Results of fit to '%s'\n" % str(function)
    notes += "   $\chi^2$ / DOF = %.2f / %d\n" \
             % (result.minimum, result.degrees_of_freedom, )
    if result.chi_square_probability is not None:
        notes += "   $P_{\chi^2}$ = %.3f\n" % result.chi_square_probability
    for name, value in result.values.items():
        notes += "   %s = %.4f ± %.4f\n" \
                 % (name, value, result.errors[name])
    result.function = hep.hist.Function1D(
        scaled_expr, arg_name=variable_name,
        axis=hep.hist.Axis(axis.type, range=axis.range), notes=notes)

    return result


def getChiSquare(histogram, function):
    """Compute the binned chi^2 between 'histogram' and 'function'.

    FIXME: Only works with one-dimensional histograms.

    returns -- 'chi_square, number_of_bins'."""

    def integrate(function, region):
        """Estimate the integral of 'function' over rectangular 'region'."""

        # Average the integrand over the corners of the region.
        sum = 0
        count = 0
        for point in hep.fn.combinations(*region):
            sum += function(*point)
            count += 1
        # Compute the volume of the region.
        volume = reduce(lambda x, (lo, hi): x * (hi - lo), region, 1)
        return volume * sum / count


    def thunk(**kw_args):
        args = []
        for n in xrange(len(kw_args)):
            args.append(kw_args["x%d" % (n, )])
        return function(*args)

    chi_square = 0
    num_bins = 0
    mult = 10 * [0]
    for bin in hep.hist.AxesIterator(histogram.axes, overflows=False):
        # Get the value and error of this histogram in this bin.
        hist_value = histogram.getBinContent(bin)
        lo, hi = histogram.getBinError(bin)
        hist_error = (lo + hi) / 2

#         # Approximate the integral of 'function' over the bin.
#         ranges = [ a.getBinRange(n)
#                    for a, n in zip(histogram.axes, bin) ]
#         # area = product([ b - a for a, b in ranges ])
#         # center = [ (a + b) / 2 for a, b in ranges ]
#         # func_value = area * function(*center)
#         integration_region = [
#             ("x%d" % (n, ), lo, hi)
#             for n, (lo, hi) in enumerate(ranges) ]
#         func_value = hep.cernlib.integrate(
#             thunk, *integration_region,
#             **{ "accuracy": 1e-4 })
        region = hep.hist.getBinRange(histogram.axes, bin)
        func_value = integrate(function, region)

        # Update the chi square.
        if func_value == 0:
            if hist_value != 0:
                print >> sys.stderr, \
                      "WARNING: skipping bin with zero function value"
        else:
            chi_square += ((func_value - hist_value) / hist_error) ** 2
            num_bins += 1
            mult[min(9, hist_value)] += 1

    print mult
    return chi_square, num_bins


def poissonLogLikelihoodRatio(histogram, function):
    """Compute the log likelihood ratio assuming Poisson bin contents.

    Computes the log likelihood ratio between a 'histogram' and a
    likelihood 'function' assuming that bin contents are Poisson
    distributed.

    'histogram' -- A histogram.

    'function' -- A likeihood function, generally produced by a (binned
    or unbinned) maximum likelihood fit."""

    def integrate(function, region):
        """Estimate the integral of 'function' over rectangular 'region'."""

        # Average the integrand over the corners of the region.
        sum = 0
        count = 0
        samples = [ ((1, l), (4, (l + h) / 2), (1, h)) for l, h in region ]
        for sample in hep.fn.combinations(*samples):
            weight = product([ w for w, c in sample ])
            point = [ c for w, c in sample ]
            sum += weight * function(*point)
            count += weight
        # Compute the volume of the region.
        volume = reduce(lambda x, (lo, hi): x * (hi - lo), region, 1)
        return volume * sum / count


    axes = histogram.axes
    log_ratio = 0
    degrees_of_freedom = 0
    for bin in hep.hist.AxesIterator(axes):
        region = hep.hist.getBinRange(axes, bin)
        f = integrate(function, region)
        h = histogram.getBinContent(bin)
        if f == 0:
            if h != 0:
                raise ValueError, \
                      "histogram nonzero where likelihood is zero"
        else:
            if h < 0:
                raise ValueError, "negative bin content"
            # print "%6d %8.1f" % (h, f)

            log_ratio += f
            if h > 0:
                log_ratio -= h * (1 + log(f) - log(h))
            degrees_of_freedom += 1

    return log_ratio, degrees_of_freedom


