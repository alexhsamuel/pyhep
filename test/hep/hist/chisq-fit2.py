#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.expr
import hep.hist
import hep.hist.fit
from   hep.test import compare, assert_
from   math import sqrt
import random

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

histogram = hep.hist.Histogram1D(20, (0.0, 4.0), bin_type=float)
contents = [ 6.8, 5.8, 4.1, 5.1, 3.9, 4.8, 4.3, 4.5, 4.0, 5.1, 5.7, 5.9,
             7.0, 8.4, 9.4, 10.7, 12.4, 14.2, 15.7, 18.0 ]
for i in range(20):
    histogram.setBinContent(i, contents[i])

expr = hep.expr.parse("a * x ** 2 + b * x + c")
result = hep.hist.fit.chiSquareFit1D(
    histogram, expr, "x", (("a", 1), ("b", 1), ("c", 1), ))
print result

compare(result.minuit_status, 3)
assert_(result.minimum < 40)
# Despite the noise, the parameter values shouldn't be *too* far off.
compare(result.values["a"],  10, precision=1)
compare(result.values["b"], -25, precision=1)
compare(result.values["c"],  35, precision=1)
