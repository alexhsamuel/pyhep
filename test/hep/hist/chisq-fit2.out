minimum value: 0.000000
parameter values:
  a                            =       1.0000 (fixed)
  b                            =       1.0000 (fixed)
  c                            =       1.0000 (fixed)
number of evaluations: 0
elapsed time: 0.01 sec
Minuit status: 0

Traceback (most recent call last):
  File "chisq-fit2.py", line 27, in ?
    compare(result.minuit_status, 3)
  File "/home/samuel/dev/pyhep/py/hep/test.py", line 59, in compare
    raise TestFailure, \
hep.test.TestFailure: expected 3; got 0
